File: apps/api/app/services/validator.py
import networkx as nx
from typing import List
from app.schemas.graph import QuestGraphPayload, ValidationErrorItem, ValidationResponse

class GraphAnalysisService:
    @classmethod
    def validate_graph(cls, graph_data: QuestGraphPayload) -> ValidationResponse:
        errors: List[ValidationErrorItem] = []
        nodes = graph_data.nodes
        edges = graph_data.edges

        if not nodes:
            return ValidationResponse(
                isValid=False,
                nodesCount=0,
                edgesCount=0,
                errors=[ValidationErrorItem(
                    severity="error",
                    code="EMPTY_GRAPH",
                    message="The graph does not contain any nodes."
                )]
            )

        # Building a directed graph in NetworkX
        G = nx.DiGraph()
        node_map = {n.id: n for n in nodes}

        for n in nodes:
            G.add_node(n.id, type=n.type, data=n.data)

        for e in edges:
            G.add_edge(e.source, e.target, id=e.id, sourceHandle=e.sourceHandle)

        # 1. Checking the root node (Root Node)
        root_id = graph_data.rootNodeId or (nodes[0].id if nodes else None)
        if not root_id or root_id not in node_map:
            errors.append(ValidationErrorItem(
                severity="error",
                code="NO_ROOT_NODE",
                message="The root node of the quest is not defined (Root Node)."
            ))

        # 2. Checking dangling edges
        for edge in edges:
            if edge.source not in node_map or edge.target not in node_map:
                errors.append(ValidationErrorItem(
                    edgeId=edge.id,
                    severity="error",
                    code="DANGLING_EDGE",
                    message=f"Edge {edge.id} points to a non-existent node ({edge.source} -> {edge.target})."
                ))

        # 3. Finding unreachable nodes (Unreachable Nodes via BFS/DFS from Root)
        if root_id and root_id in G:
            reachable = nx.descendants(G, root_id)
            reachable.add(root_id)

            for node_id in node_map.keys():
                if node_id not in reachable:
                    errors.append(ValidationErrorItem(
                        nodeId=node_id,
                        severity="warning",
                        code="UNREACHABLE_NODE",
                        message=f"Node [{node_map[node_id].data.get('label', node_id)}] is unreachable from the starting point."
                    ))

        # 4. Finding infinite cycles without exit options
        try:
            cycles = list(nx.simple_cycles(G))
            for cycle in cycles:
                cycle_names = [node_map[nid].data.get('label', nid) for nid in cycle if nid in node_map]
                errors.append(ValidationErrorItem(
                    nodeId=cycle[0],
                    severity="warning",
                    code="DETECTED_CYCLE",
                    message=f"Found a cycle: {' -> '.join(cycle_names)}"
                ))
        except Exception:
            pass

        # 5. Checking dialogue nodes without response options
        for n in nodes:
            if n.type == "dialogueNode":
                options = n.data.get("options", [])
                if not options:
                    errors.append(ValidationErrorItem(
                        nodeId=n.id,
                        severity="warning",
                        code="EMPTY_CHOICES",
                        message=f"Dialogue [{n.data.get('speakerName', 'NPC')}] does not contain response options."
                    ))

        has_errors = any(e.severity == "error" for e in errors)

        return ValidationResponse(
            isValid=not has_errors,
            nodesCount=len(nodes),
            edgesCount=len(edges),
            errors=errors
        )

File: lib/engine/validator.ts
import { QuestGraph, GraphValidationResult, GraphValidationError, DialogueNodeData, AnyNodeData } from '@/types';

export class GraphValidator {
    public static validate(graph: QuestGraph): GraphValidationResult {
        const errors: GraphValidationError[] = [];
        const nodes = (graph.nodes || []).filter(Boolean);
        const edges = (graph.edges || []).filter(Boolean);

        if (nodes.length === 0) {
            errors.push({
                severity: 'error',
                code: 'EMPTY_GRAPH',
                message: 'The graph does not contain any nodes.',
            });
            return {
                isValid: false,
                errors,
                nodesCount: 0,
                edgesCount: 0,
            };
        }

        const nodeMap = new Map(nodes.map((n) => [n.id, n]));
        const rootId = graph.rootNodeId || nodes[0]?.id;

        // 1. Checking starting node
        if (!rootId || !nodeMap.has(rootId)) {
            errors.push({
                severity: 'error',
                code: 'NO_ROOT_NODE',
                message: `Root node '${rootId}' is not found in the graph.`,
            });
        }

        // 2. Checking dangling edges
        for (const edge of edges) {
            if (!nodeMap.has(edge.source)) {
                errors.push({
                    edgeId: edge.id,
                    severity: 'error',
                    code: 'DANGLING_EDGE',
                    message: `Edge [${edge.id}] has non-existent source: ${edge.source}`,
                });
            }
            if (!nodeMap.has(edge.target)) {
                errors.push({
                    edgeId: edge.id,
                    severity: 'error',
                    code: 'DANGLING_EDGE',
                    message: `Edge [${edge.id}] has non-existent target: ${edge.target}`,
                });
            }
        }

        // Building adjacency list
        const adj = new Map<string, string[]>();
        for (const node of nodes) {
            adj.set(node.id, []);
        }
        for (const edge of edges) {
            if (adj.has(edge.source)) {
                adj.get(edge.source)!.push(edge.target);
            }
        }

        // 3. Checking node reachability from root
        if (rootId && nodeMap.has(rootId)) {
            const reachable = new Set<string>();
            const queue = [rootId];
            reachable.add(rootId);

            while (queue.length > 0) {
                const curr = queue.shift()!;
                for (const next of adj.get(curr) || []) {
                    if (!reachable.has(next) && nodeMap.has(next)) {
                        reachable.add(next);
                        queue.push(next);
                    }
                }
            }

            for (const node of nodes) {
                if (!reachable.has(node.id)) {
                    const label = (node.data as AnyNodeData)?.label || node.id;
                    errors.push({
                        nodeId: node.id,
                        severity: 'warning',
                        code: 'UNREACHABLE_NODE',
                        message: `Node [${label}] is unreachable from the starting point.`,
                    });
                }
            }
        }

        // 4. Finding cycles (DFS cycle detection)
        const cycles = this.findSimpleCycles(adj);
        for (const cycle of cycles) {
            const cycleLabels = cycle.map((id) => (nodeMap.get(id)?.data as AnyNodeData)?.label || id);
            errors.push({
                nodeId: cycle[0],
                severity: 'warning',
                code: 'DETECTED_CYCLE',
                message: `Found a cycle: ${cycleLabels.join(' -> ')}`,
            });
        }

        // 5. Checking dialogue nodes without response options
        for (const node of nodes) {
            if (node.type === 'dialogue' || node.type === ('dialogueNode' as string)) {
                const options = (node.data as DialogueNodeData)?.options;
                if (!options || options.length === 0) {
                    const speaker = (node.data as DialogueNodeData)?.speakerName || 'NPC';
                    errors.push({
                        nodeId: node.id,
                        severity: 'warning',
                        code: 'EMPTY_CHOICES',
                        message: `Dialogue [${speaker}] does not contain response options.`,
                    });
                }
            }
        }

        const hasErrors = errors.some((e) => e.severity === 'error');

        return {
            isValid: !hasErrors,
            nodesCount: nodes.length,
            edgesCount: edges.length,
            errors,
        };
    }

    /**
     * Find elementary directed cycles using DFS
     */
    private static findSimpleCycles(adj: Map<string, string[]>): string[][] {
        const cycles: string[][] = [];
        const visitedState = new Map<string, number>(); // 0: unvisited, 1: visiting, 2: visited
        const currentPath: string[] = [];
        const reportedCycles = new Set<string>();

        const dfs = (node: string) => {
            visitedState.set(node, 1);
            currentPath.push(node);

            const neighbors = adj.get(node) || [];
            for (const neighbor of neighbors) {
                const state = visitedState.get(neighbor) || 0;
                if (state === 1) {
                    // Cycle detected! Extract cycle from neighbor to end of currentPath
                    const cycleStartIndex = currentPath.indexOf(neighbor);
                    if (cycleStartIndex !== -1) {
                        const cycle = currentPath.slice(cycleStartIndex);
                        const cycleKey = [...cycle].sort().join(',');
                        if (!reportedCycles.has(cycleKey)) {
                            reportedCycles.add(cycleKey);
                            cycles.push([...cycle, neighbor]);
                        }
                    }
                } else if (state === 0) {
                    dfs(neighbor);
                }
            }

            currentPath.pop();
            visitedState.set(node, 2);
        };

        for (const node of adj.keys()) {
            if ((visitedState.get(node) || 0) === 0) {
                dfs(node);
            }
        }

        return cycles;
    }
}